README.md
# paperless
A project to help me go paperless

App to go paperless.
Backend using Flask. Connects to local scanner and gives access to:
- scan?resolution (to temp file. redirects to view)
- view?id (view temp file)
- delete?id  (delete temp file)
- save?id (save in spideroak directory. if not path, to incoming)
- add?id&file (add id to file (pdf). If no file, generate and return)

Frontend having multiple tabs:
 - config (set API ip, resolution)
 - scan page
   - scan button(75 dpi)
   - preview image
   - delete button
   - save button (rescan with configured dpi)
   - add button (to scan another page and add it to a pdf)

HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2015-04-08)
---------------------

* First release on pypi.office.comscore.com.


lib/main.js
/**
 * Created by wfournier on 4/12/15.
 */
var api = 'http://192.168.54.106:5000';
//var api = 'http://localhost:5000';
var image = '';
var destinations = [];
var resolution = 300;

function show_notification(text) {
    console.log("Notification called with: " + text);
    var dialog = $("#dialog");
        //.find("p").text(text)
        dialog.html(
            "<p>" + text + "</p>"
        )
        .attr("class", "notification")
        .attr("heading", "Notification")
        .toggle();
    return dialog;
}

function show_error(text) {
    console.log("Error called with: " + text);
    var dialog = $("#dialog");
        //.find("p").text(text)
        dialog.html(
            "<p>" + text + "</p>" +
            "<paper-button affirmative=\"\" autofocus=\"\" role=\"button\" tabindex=\"0\">Ok</paper-button>"
        )
        .attr("class", "error")
        .attr("heading", "Error")
        .toggle()
        .find("paper-button").click(function() {
            $(this).parent().toggle();
        });
    return dialog;
}

function scan() {
    console.log("Scan called");
    var notification = show_notification("Please wait while the scanner is scanning...");
    $.get(api + '/scan?resolution=' + resolution, function(data) {
    //$.get(api + '/test', function(data) {
        console.log(data);
        image = data;
        $("#preview").attr("src", api + "/view?id=" + data);
        $("#save")
            .removeAttr("disabled")
            .removeAttr("aria-disabled")
            .attr("class", "colored");
    })
        .done(function(data) {
            console.log('Scan.done called');
            notification.toggle();
            $("#scan").attr("class", "succeeded");
            setTimeout(function(){
                $("#scan").attr("class", "colored");
            }, 2000);
        })
        .fail(function(jqXHR, textStatus, errorThrown) {
            console.log('Scan.fail called');
            notification.toggle();
            $("#scan").attr("class", "failed");
            show_error("Scan failed with: " + textStatus + "\n" + errorThrown + "\n" + jqXHR.responseText);

        });
}

function save_dialog() {
    document.querySelector("#save-dialog").toggle_dialog();
}

function save() {
    console.log('Save called');
    document.querySelector("#save-dialog").toggle_dialog();
    var location = $(".core-selected").text();
    if (location == '') {
        show_error("Please select a destination first...");
        return false;
    }
    var file_name = document.querySelector('#save-dialog').file_name;
    console.log('file_name :' + file_name);
    if (file_name == '') {
        console.log('file_name empty');
        show_error("Please put in a file_name...");
        return false;
    }
    $.get(api + "/save?id=" + image + "&location=" + location + "&name=" + file_name,
        function(data) {})
        .done(function(data) {
            console.log('Save ok');
            $("#save").attr("class", "succeeded");
            $("#preview").attr("src", "");
            setTimeout(function(){
                $("#save")
                    .attr("disabled", "")
                    .attr("aria-disabled","")
                    .removeAttr("class");
            }, 2000);
        })
        .fail(function(jqXHR, textStatus, errorThrown) {
            console.log('Save failed');
            $("#save").attr("class", "failed");
            show_error("Save failed with: " + textStatus + "\n" + errorThrown + "\n" + jqXHR.responseText);
        });
}

function populate_locations() {
    console.log('populate_locations called');
    $.get(api + '/destinations', function(data) {
        console.log(data);
        destinations = JSON.parse(data).sort();
        for (var i in destinations) {
            $("#destination_list").append("<paper-item>"+ destinations[i] +"</paper-item>");
        }
    })
        .fail(function(jqXHR, textStatus, errorThrown) {
            console.log('populate_locations failed');
            show_error("Populate of locations failed with: " + textStatus + "\n" + errorThrown + "\n" + jqXHR.responseText);
        });
}


paperless/app.py
#!/usr/bin/env python2.7
# -*- coding: UTF-8 -*-
# File: paperless.py

#
__author__ = 'Wim Fournier <wim@fournier.nl>'
__docformat__ = 'plaintext'
__date__ = '2015-04-08'

from flask import Flask, redirect, url_for, Response, request
import pyinsane.abstract as pyinsane
from tempfile import NamedTemporaryFile
import os
import json
import shutil
from datetime import datetime
import time

app = Flask(__name__)
images = [{'path': '/tmp/tmpLOKWW9', 'id': 'tmpLOKWW9'}]
base_path = '/home/wfournier/SpiderOak Hive/paperless'


# @app.route('/scan')
def test():
    time.sleep(1)
    return 'tmpLOKWW9', 200


@app.route('/destinations')
def destinations():
    directories = []
    for root, dirs, files in os.walk(base_path):
        directories.append(root[len(base_path) + 1:])
    return json.dumps(directories), 200


@app.route('/save')
def save():
    # return 'OK', 200
    image_id = request.args.get('id', '')
    location = request.args.get('location', '')
    name = request.args.get('name', '')
    file_name = '{datetime} {name}.jpg'.format(
        datetime=datetime.strftime(datetime.now(), '%Y%m%d-%H%M%S'),
        name=name)
    if not id or not location:
        return 'Missing id or location', 500
    image = find_image(image_id)
    if image:
        try:
            destination_path = os.path.join(base_path, location, file_name)
            print 'Saving {} to {}'.format(image['path'], destination_path)
            shutil.move(image['path'], destination_path)
        except shutil.Error as e:
            print e
            return str(e), 500
        return 'OK', 200
    else:
        return 'Image not found', 404


@app.route('/scan')
def scan():
    resolution = request.args.get('resolution', '')
    print 'resolution: {}'.format(resolution)
    devices = pyinsane.get_devices()
    assert(len(devices) > 0)
    device = devices[0]
    if resolution and resolution in ['75', '150', '300', '600', '1200', '2400']:
        device.options['resolution'].value = int(resolution)
    print device.options['resolution'].value
    print device.name
    # for option in device.options:
    #     try:
    #         print option, device.options[option].value
    #     except:
    #         print option
    scan_session = device.scan(multiple=False)
    try:
        while True:
            print '.'
            scan_session.scan.read()
    except EOFError:
        pass
    print 'scan done'

    image = scan_session.images[0]
    tmp_file = NamedTemporaryFile(delete=False)
    image.save(tmp_file.name, "JPEG")
    image_id = tmp_file.name.rpartition('/')[-1]
    images.append({'id': image_id, 'path': tmp_file.name})
    return image_id, 200


def find_image(image_id):
    image = ''
    for item in images:
        if item['id'] == image_id:
            print 'found image at {}'.format(item['path'])
            image = item
    return image


@app.route('/view')
def view():
    print images
    image_id = request.args.get('id', '')
    if not image_id:
        return 'No image ID given', 500
    image = find_image(image_id)
    if image:
        return Response(open(image['path']).read(), mimetype='image/jpeg')
    else:
        return 'Image not found', 404


@app.route('/delete')
def delete():
    image_id = request.args.get('id', '')
    if not image_id:
        return 'No image ID given', 500
    for item in images:
        if item['id'] == image_id:
            print 'found image at {}'.format(item['path'])
            os.remove(item['path'])
            return '{} has been removed'.format(image_id), 200
    return 'Image not found', 404


@app.after_request
def add_cors(resp):
    """ Ensure all responses have the CORS headers. This ensures any failures are also accessible
        by the client. """
    resp.headers['Access-Control-Allow-Origin'] = request.headers.get('Origin','*')
    resp.headers['Access-Control-Allow-Credentials'] = 'true'
    resp.headers['Access-Control-Allow-Methods'] = 'POST, OPTIONS, GET'
    resp.headers['Access-Control-Allow-Headers'] = request.headers.get(
        'Access-Control-Request-Headers', 'Authorization' )
    # set low for debugging
    if app.debug:
        resp.headers['Access-Control-Max-Age'] = '1'
    return resp


if __name__ == '__main__':
    app.run(debug=True, host='0.0.0.0')
